samples/extensions/dynamic_rendering/dynamic_rendering.cpp

```diff
@@ -0,0 +1,466 @@
+
+void DynamicRendering::draw()
+{
+	ApiVulkanSample::prepare_frame();
+	submit_info.commandBufferCount = 1;
+	submit_info.pCommandBuffers    = &draw_cmd_buffers[current_buffer];
+	VK_CHECK(vkQueueSubmit(queue, 1, &submit_info, VK_NULL_HANDLE));
+	ApiVulkanSample::submit_frame();
+}
+
+void DynamicRendering::build_command_buffers()
+{
+	std::array<VkClearValue, 2> clear_values{};
+	clear_values[0].color        = {{0.0f, 0.0f, 0.0f, 0.0f}};
+	clear_values[1].depthStencil = {0.0f, 0};
+
+	int i = -1;
+	for (auto &draw_cmd_buffer : draw_cmd_buffers)
+	{
+		i++;
+		auto command_begin = vkb::initializers::command_buffer_begin_info();
+		VK_CHECK(vkBeginCommandBuffer(draw_cmd_buffer, &command_begin));
+
+		auto draw_scene = [&] {
+			VkViewport viewport = vkb::initializers::viewport((float) width, (float) height, 0.0f, 1.0f);
+			vkCmdSetViewport(draw_cmd_buffer, 0, 1, &viewport);
+
+			VkRect2D scissor = vkb::initializers::rect2D(static_cast<int>(width), static_cast<int>(height), 0, 0);
+			vkCmdSetScissor(draw_cmd_buffer, 0, 1, &scissor);
+
+			// One descriptor set is used, and the draw type is toggled by a specialization constant
+			vkCmdBindDescriptorSets(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, pipeline_layout, 0, 1, &descriptor_set, 0, nullptr);
+
+			// skybox
+			vkCmdBindPipeline(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, skybox_pipeline);
+			draw_model(skybox, draw_cmd_buffer);
+
+			// object
+			vkCmdBindPipeline(draw_cmd_buffer, VK_PIPELINE_BIND_POINT_GRAPHICS, model_pipeline);
+			draw_model(object, draw_cmd_buffer);
+
+			// UI
+			draw_ui(draw_cmd_buffer);
+		};
+
+		VkImageSubresourceRange range{};
+		range.aspectMask     = VK_IMAGE_ASPECT_COLOR_BIT;
+		range.baseMipLevel   = 0;
+		range.levelCount     = VK_REMAINING_MIP_LEVELS;
+		range.baseArrayLayer = 0;
+		range.layerCount     = VK_REMAINING_ARRAY_LAYERS;
+
+		VkImageSubresourceRange depth_range{range};
+		depth_range.aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
+
+		if (enable_dynamic)
+		{
+			vkb::insert_image_memory_barrier(draw_cmd_buffer,
+			                                 swapchain_buffers[i].image,
+			                                 0,
+			                                 VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
+			                                 VK_IMAGE_LAYOUT_UNDEFINED,
+			                                 VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+			                                 VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
+			                                 VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT, range);
+
+			vkb::insert_image_memory_barrier(draw_cmd_buffer,
+			                                 depth_stencil.image,
+			                                 0,
+			                                 VK_ACCESS_DEPTH_STENCIL_ATTACHMENT_WRITE_BIT,
+			                                 VK_IMAGE_LAYOUT_UNDEFINED,
+			                                 VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_OPTIMAL,
+			                                 VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT,
```

Comment:
Common case is last use was `VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT` in previous frame, so that seems correct stage here.